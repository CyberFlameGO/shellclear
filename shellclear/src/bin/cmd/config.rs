use anyhow::Result;
use clap::{Arg, ArgMatches, Command};
use shellclear::config::Config;
use shellclear::promter;

pub fn command() -> Command<'static> {
    Command::new("config")
        .about("Create custom configuration")
        .arg(
            Arg::new("validate")
                .short('v')
                .long("validate")
                .help("Validate configuration file.")
                .takes_value(false),
        )
        .arg(
            Arg::new("delete")
                .short('d')
                .long("delete")
                .help("Delete configuration file.")
                .takes_value(false),
        )
}

pub fn run(matches: &ArgMatches, config: &Config) -> Result<shellclear::CmdExit> {
    let file_path = config.get_sensitive_pattern_name();

    let (message, exit_code) = if matches.is_present("validate") {
        if let Err(e) = config.load_patterns_from_default_path() {
            (
                format!("Config file is invalid. error: `{}`.", e),
                exitcode::CONFIG,
            )
        } else {
            (
                format!("configuration file: {} is valid.", file_path),
                exitcode::OK,
            )
        }
    } else if matches.is_present("delete") {
        if let Err(e) = promter::confirm(format!("Delete {} file?", &file_path).as_str()) {
            log::debug!("{:?}", e);
            return Ok(shellclear::CmdExit {
                code: exitcode::OK,
                message: None,
            });
        }

        if config.is_sensitive_pattern_file_exists() {
            config.delete_sensitive_patterns_from_file()?;
        }

        (
            format!("Config file deleted successfully in path: {}", file_path),
            exitcode::OK,
        )
    } else {
        if config.is_sensitive_pattern_file_exists() {
            let confirm_message = format!(
                "file {} already exists. do you want to override the existing file?",
                file_path,
            );
            if let Err(e) = promter::confirm(&confirm_message) {
                log::debug!("{:?}", e);
                return Ok(shellclear::CmdExit {
                    code: exitcode::OK,
                    message: None,
                });
            }
        }
        config.init()?;
        (
            format!("Config file created successfully in path: {}", file_path),
            exitcode::OK,
        )
    };

    Ok(shellclear::CmdExit {
        code: exit_code,
        message: Some(message),
    })
}
