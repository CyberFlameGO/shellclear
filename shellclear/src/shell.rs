use anyhow::anyhow;
use anyhow::Result;
use std::fmt;
use std::path::Path;
use strum::{EnumIter, IntoEnumIterator};

/// List of all supported shells
#[derive(Debug, EnumIter, Clone)]
pub enum Shell {
    Bash,
    Zshrc,
    Fish,
}

impl fmt::Display for Shell {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Self::Bash => write!(f, "bash"),
            Self::Zshrc => write!(f, "zshrc"),
            Self::Fish => write!(f, "fish"),
        }
    }
}

/// Zsh history file name
const ZSH_HISTORY_FILE_PATH: &str = ".zsh_history";
/// Bash history file name
const BASH_HISTORY_FILE_PATH: &str = ".bash_history";
const FISH_HISTORY_FILE_PATH: &str = ".local/share/fish/fish_history";

/// History shell details
#[derive(Clone, Debug)]
pub struct History {
    /// shell type
    pub shell: Shell,
    /// history file path
    pub path: String,
    /// history file name
    pub file_name: String,
}

/// return list of all existing history files
pub fn get_all_history_files() -> Result<Vec<History>> {
    let homedir = match dirs::home_dir() {
        Some(h) => h.display().to_string(),
        None => return Err(anyhow!("could not get directory path")),
    };

    // return list of existing shells
    Ok(Shell::iter()
        .map(|shell| {
            let shell_history_path = get_shell_history_path(&shell, &homedir);
            if !Path::new(&shell_history_path).exists() {
                log::debug!("shell {:?} not found", shell);
                return Err(anyhow!("could not get directory path"));
            }
            let file_name = Path::new(&shell_history_path)
                .file_name()
                .unwrap()
                .to_str()
                .unwrap()
                .to_string();
            Ok(History {
                shell,
                path: shell_history_path,
                file_name,
            })
        })
        .filter(std::result::Result::is_ok)
        .map(std::result::Result::unwrap)
        .collect::<Vec<_>>())
}

// returns all supported shells types
fn get_shell_history_path(shell_type: &Shell, homedir: &str) -> String {
    match shell_type {
        Shell::Bash => Path::new(homedir)
            .join(BASH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
        Shell::Zshrc => Path::new(homedir)
            .join(ZSH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
        Shell::Fish => Path::new(homedir)
            .join(FISH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
    }
}
