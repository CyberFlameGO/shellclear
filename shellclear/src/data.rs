use crate::shell::Shell;
use serde_derive::Deserialize;

pub struct CmdExit {
    pub code: exitcode::ExitCode,
    pub message: Option<String>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct SensitiveCommands {
    #[serde(with = "serde_regex")]
    pub test: regex::Regex,
    pub name: String,
}

#[derive(Debug)]
pub struct FindingSensitiveCommands {
    pub shell_type: Shell,
    pub sensitive_findings: Vec<SensitiveCommands>,
    pub command: String,
}
