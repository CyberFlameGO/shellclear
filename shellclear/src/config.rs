use crate::data::{SensitiveCommands, ROOT_APP_FOLDER};
use anyhow::Result;
use std::fs;
use std::path::PathBuf;

const CONFIG_SENSITIVE_PATTERNS: &str = "sensitive-patterns.yaml";
const SENSITIVE_PATTERN_TEMPLATE: &str = r###"# External sensitive patters file allows you you add a custom patterns to shellclear

- name: Pattern Name
  test: <PATTERN REGEX>
"###;

#[derive(Clone, Debug)]
pub struct Config {
    sensitive_commands_path: PathBuf,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            sensitive_commands_path: Config::get_sensitive_pattern_file(&dirs::home_dir().unwrap()),
        }
    }
}

impl Config {
    fn with_custom_path(root: &PathBuf) -> Self {
        Self {
            sensitive_commands_path: Self::get_sensitive_pattern_file(root),
        }
    }
    /// Init external configuration
    ///
    /// # Errors
    ///
    /// Will return `Err` when home directory not found or failed to create a file
    pub fn init(&self) -> Result<String> {
        let file = &self.sensitive_commands_path;
        fs::write(file, SENSITIVE_PATTERN_TEMPLATE)?;
        Ok(file.display().to_string())
    }

    /// Get sensitive file path name
    ///
    /// # Errors
    ///
    /// Will return `Err` home directory not found
    #[must_use]
    pub fn get_sensitive_pattern_name(&self) -> String {
        self.sensitive_commands_path.display().to_string()
    }

    /// Is sensitive file is exists
    ///
    /// # Errors
    ///
    /// Will return `Err` home directory not found
    #[must_use]
    pub fn is_sensitive_pattern_file_exists(&self) -> bool {
        self.sensitive_commands_path.exists()
    }

    /// Load sensitive pattern file
    ///
    /// # Errors
    ///
    /// Will return `Err` home directory not found or yaml is invalid
    pub fn load_patterns_from_default_path(&self) -> Result<Vec<SensitiveCommands>> {
        self.load_patterns_from_file(&self.sensitive_commands_path)
    }

    /// Load sensitive pattern from the given path
    ///
    /// # Errors
    ///
    /// Will return `Err`  yaml is invalid
    fn load_patterns_from_file(&self, path: &PathBuf) -> Result<Vec<SensitiveCommands>> {
        let f = std::fs::File::open(path)?;
        Ok(serde_yaml::from_reader(f)?)
    }

    /// Returns the root shellclear config folder
    fn get_sensitive_pattern_file(path: &PathBuf) -> PathBuf {
        path.join(format!(".{}", ROOT_APP_FOLDER))
            .join(CONFIG_SENSITIVE_PATTERNS)
    }
}
