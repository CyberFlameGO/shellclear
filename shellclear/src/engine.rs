use crate::config::Config;
use crate::data::{FindingSensitiveCommands, SensitiveCommands};
use crate::shell;
use crate::state::ShellContext;
use anyhow::Result;
use log::debug;
use rayon::prelude::*;
use serde_derive::{Deserialize, Serialize};
use std::fmt::Write;
use std::fs::{write, File};
use std::io::{prelude::*, BufReader};
use std::time::Instant;

pub const SENSITIVE_COMMANDS: &str = include_str!("sensitive-patterns.yaml");

#[derive(Debug, Deserialize, Serialize, Clone)]
struct FishHistory {
    pub cmd: String,
    pub when: String,
}

pub struct PatternsEngine {
    sensitive_commands: Vec<SensitiveCommands>,
}

#[derive(Default)]
pub struct Findings {
    pub patterns: Vec<FindingSensitiveCommands>,
}

impl Default for PatternsEngine {
    fn default() -> Self {
        let sensitive_patterns = {
            let mut patterns: Vec<SensitiveCommands> =
                serde_yaml::from_str(SENSITIVE_COMMANDS).unwrap();
            match Config::load_patterns_from_default_path() {
                Ok(p) => patterns.extend(p),
                Err(e) => log::debug!("{:?}", e),
            };
            patterns
        };
        Self {
            sensitive_commands: sensitive_patterns,
        }
    }
}

impl Findings {
    #[must_use]
    // add list of finding
    pub fn add_findings(mut self, finding: Vec<FindingSensitiveCommands>) -> Self {
        self.patterns.extend(finding);
        self
    }

    #[must_use]
    // return list of sensitive findings commands
    pub fn get_sensitive_commands(&self) -> Vec<&FindingSensitiveCommands> {
        self.patterns
            .iter()
            .filter(|&f| !f.sensitive_findings.is_empty())
            .collect::<Vec<_>>()
    }
}

impl PatternsEngine {
    /// Search sensitive command patterns from the given sehll list
    ///
    /// # Errors
    ///
    /// Will return `Err` when has an error when find sensitive patters in a specific shell
    pub fn find_history_commands_from_shall_list(
        &self,
        shells_context: &Vec<ShellContext>,
        clear: bool,
    ) -> Result<Findings> {
        let mut findings = Findings::default();

        for shell_context in shells_context {
            findings = findings.add_findings(self.find_history_commands(shell_context, clear)?);
        }
        Ok(findings)
    }
    /// Search sensitive command patterns
    ///
    /// # Errors
    ///
    /// Will return `Err` if history file not exists/ could't open
    pub fn find_history_commands(
        &self,
        state_context: &ShellContext,
        clear: bool,
    ) -> Result<Vec<FindingSensitiveCommands>> {
        debug!(
            "clear history commands from path: {}, params: is clear: {}",
            state_context.history.path, clear
        );

        match state_context.history.shell {
            shell::Shell::Fish => self.find_fish(state_context, &self.sensitive_commands, clear),
            _ => self.find_by_lines(state_context, &self.sensitive_commands, clear),
        }
    }

    fn find_by_lines(
        &self,
        state_context: &ShellContext,
        sensitive_commands: &Vec<SensitiveCommands>,
        clear: bool,
    ) -> Result<Vec<FindingSensitiveCommands>> {
        let file = File::open(&state_context.history.path)?;
        let reader = BufReader::new(file);

        let start = Instant::now();

        let lines = reader
            .lines()
            .filter(std::result::Result::is_ok)
            .map(std::result::Result::unwrap)
            .collect::<Vec<_>>();

        debug!(
            "time elapsed to read history file: {:?}. found {} commands",
            start.elapsed(),
            lines.len()
        );

        let start = Instant::now();
        let results = lines
            .par_iter()
            .map(|command| {
                let sensitive_findings = sensitive_commands
                    .par_iter()
                    .filter(|v| v.test.is_match(command))
                    .map(std::clone::Clone::clone)
                    .collect::<Vec<_>>();

                FindingSensitiveCommands {
                    shell_type: state_context.history.shell.clone(),
                    sensitive_findings,
                    command: command.clone(),
                }
            })
            .collect::<Vec<_>>();

        debug!(
            "time elapsed for detect sensitive commands: {:?}",
            start.elapsed()
        );

        if clear {
            let start = Instant::now();
            let mut cleared_history: String = String::new();

            for r in &results {
                if r.sensitive_findings.is_empty() {
                    let _ = writeln!(&mut cleared_history, "{}", r.command);
                }
            }
            if !cleared_history.is_empty() {
                write(&state_context.history.path, cleared_history)?;
                debug!(
                    "time elapsed for backup existing file and write a new history to shell : {:?}",
                    start.elapsed()
                );
            }
        }

        Ok(results)
    }

    fn find_fish(
        &self,
        state_context: &ShellContext,
        sensitive_commands: &Vec<SensitiveCommands>,
        clear: bool,
    ) -> Result<Vec<FindingSensitiveCommands>> {
        let start = Instant::now();
        let history: Vec<FishHistory> =
            serde_yaml::from_reader(File::open(&state_context.history.path)?)?;

        let results = history
            .par_iter()
            .map(|h| {
                let sensitive_findings = sensitive_commands
                    .par_iter()
                    .filter(|v| v.test.is_match(&h.cmd))
                    .map(std::clone::Clone::clone)
                    .collect::<Vec<_>>();

                FindingSensitiveCommands {
                    shell_type: state_context.history.shell.clone(),
                    sensitive_findings,
                    command: serde_yaml::to_string(&h).unwrap(),
                }
            })
            .collect::<Vec<_>>();
        debug!(
            "time elapsed to read history file: {:?}. found {} commands",
            start.elapsed(),
            history.len()
        );

        if clear {
            let start = Instant::now();
            let mut cleared_history: Vec<FishHistory> = Vec::new();

            for command_line in &results {
                if command_line.sensitive_findings.is_empty() {
                    cleared_history.push(serde_yaml::from_str(&command_line.command)?);
                }
            }
            if !cleared_history.is_empty() {
                write(
                    &state_context.history.path,
                    serde_yaml::to_string(&cleared_history).unwrap(),
                )?;
                debug!(
                    "time elapsed for backup existing file and write a new history to shell : {:?}",
                    start.elapsed()
                );
            }
        }
        Ok(results)
    }
}

#[cfg(test)]
mod test_engine {
    use super::*;
    use insta::assert_debug_snapshot;
    use regex::Regex;
    use std::fs;
    use std::fs::File;
    use std::io::Write;
    use tempdir::TempDir;

    const TEMP_HISTORY_LINES_CONTENT: &str = "history
ls
echo 'hello you'
rm -f ./file.txt
export DELETE_ME=token
";

    const TEMP_HISTORY_FISH: &str = r#"---
- cmd: history
  when: "1656438759"
- cmd: ls
  when: "1656438760"
- cmd: echo 'hello you'
  when: "1656438760"
- cmd: rm -f ./file.txt
  when: "1656438760"
- cmd: export DELETE_ME=token
  when: "1656438760"
"#;

    fn create_mock_state(
        temp_dir: &TempDir,
        content: &str,
        shell_type: shell::Shell,
    ) -> ShellContext {
        let app_folder = temp_dir.path().join("app");
        let history_file_name = "history";
        let history_file_path = app_folder.join(history_file_name);
        fs::create_dir_all(&app_folder).unwrap();

        let mut f = File::create(&history_file_path).unwrap();
        f.write_all(content.as_bytes()).unwrap();
        f.sync_all().unwrap();

        ShellContext {
            app_folder_path: app_folder.display().to_string(),
            history: shell::History {
                shell: shell_type,
                path: history_file_path.display().to_string(),
                file_name: history_file_name.to_string(),
            },
        }
    }

    #[test]
    fn can_find_history_commands_line() {
        let temp_dir = TempDir::new("engine").unwrap();

        let en = PatternsEngine::default();
        let state_context =
            create_mock_state(&temp_dir, TEMP_HISTORY_LINES_CONTENT, shell::Shell::Bash);

        let result = en.find_history_commands(&state_context, false);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn can_clear_command_by_lines() {
        let temp_dir = TempDir::new("engine").unwrap();
        let search_sensitive_commands = vec![SensitiveCommands {
            name: "test".to_string(),
            test: Regex::new("DELETE_ME").unwrap(),
        }];

        let en = PatternsEngine::default();
        let state_context =
            create_mock_state(&temp_dir, TEMP_HISTORY_LINES_CONTENT, shell::Shell::Bash);

        let result = en.find_by_lines(&state_context, &search_sensitive_commands, true);

        assert_debug_snapshot!(result);
        assert_debug_snapshot!(fs::read_to_string(state_context.history.path));
    }

    #[test]
    fn can_clear_find_fish() {
        let temp_dir = TempDir::new("engine").unwrap();
        let search_sensitive_commands = vec![SensitiveCommands {
            name: "test".to_string(),
            test: Regex::new("DELETE_ME").unwrap(),
        }];

        let en = PatternsEngine::default();
        let state_context = create_mock_state(&temp_dir, TEMP_HISTORY_FISH, shell::Shell::Fish);
        let result = en.find_fish(&state_context, &search_sensitive_commands, true);

        assert_debug_snapshot!(result);
        assert_debug_snapshot!(fs::read_to_string(state_context.history.path).unwrap());
    }

    #[test]
    fn can_find_history_commands_fish() {
        let temp_dir = TempDir::new("engine").unwrap();

        let en = PatternsEngine::default();
        let state_context = create_mock_state(&temp_dir, TEMP_HISTORY_FISH, shell::Shell::Fish);

        let result = en.find_history_commands(&state_context, false);

        assert_debug_snapshot!(result);
    }
}
