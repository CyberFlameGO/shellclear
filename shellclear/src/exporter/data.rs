use crate::data::FindingSensitiveCommands;
use crate::shell::Shell;
use anyhow::Result;
use chrono::{DateTime, Local, NaiveDateTime, TimeZone};
use lazy_static::lazy_static;
use regex::Regex;

const DATE_TIME_FORMAT: &str = "%Y-%m-%d %H:%M:%S";

pub trait Exporter {
    fn sensitive_data(&self, findings: &[&FindingSensitiveCommands]) -> Result<()>;
}

lazy_static! {
    static ref ZSHRC_CAPTURE_TIME: Regex = Regex::new(r"^: ([0-9]+):").unwrap();
}

// todo:: move under export trait
pub fn chunk(text: &str, size: usize) -> String {
    text.chars()
        .collect::<Vec<char>>()
        .chunks(size)
        .map(|c| c.iter().collect::<String>())
        .collect::<Vec<String>>()
        .join("\r\n")
}

/// Get command time execution
///
/// # Errors
///
/// Will return `Err` when conversion error
pub fn extract_time(finding: &FindingSensitiveCommands) -> Result<String> {
    match finding.shell_type {
        Shell::Zshrc => {
            if let Some(c) = ZSHRC_CAPTURE_TIME.captures(&finding.data) {
                if let Some(timestemp) = c.get(1) {
                    let date_time: DateTime<Local> = Local.from_utc_datetime(
                        &NaiveDateTime::from_timestamp(timestemp.as_str().parse::<i64>()?, 0),
                    );
                    return Ok(format!("{}", date_time.format(DATE_TIME_FORMAT)));
                }
            };
            Ok(finding.data.clone())
        }
        _ => Ok("".to_string()),
    }
}
