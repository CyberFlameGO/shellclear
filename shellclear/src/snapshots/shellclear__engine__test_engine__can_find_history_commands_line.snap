---
source: shellclear/src/engine.rs
expression: result
---
Ok(
    [
        FindingSensitiveCommands {
            shell_type: Bash,
            sensitive_findings: [],
            command: "history",
        },
        FindingSensitiveCommands {
            shell_type: Bash,
            sensitive_findings: [],
            command: "ls",
        },
        FindingSensitiveCommands {
            shell_type: Bash,
            sensitive_findings: [],
            command: "echo 'hello you'",
        },
        FindingSensitiveCommands {
            shell_type: Bash,
            sensitive_findings: [],
            command: "rm -f ./file.txt",
        },
        FindingSensitiveCommands {
            shell_type: Bash,
            sensitive_findings: [],
            command: "export DELETE_ME=token",
        },
    ],
)
